crypto/bn/bn_mont.c

```diff
@@ -20,29 +20,43 @@
 #define MONT_WORD               /* use the faster word-based algorithm */
 
 #ifdef MONT_WORD
-static int BN_from_montgomery_word(BIGNUM *ret, BIGNUM *r, BN_MONT_CTX *mont);
+static int bn_from_montgomery_word(BIGNUM *ret, BIGNUM *r, BN_MONT_CTX *mont);
 #endif
 
 int BN_mod_mul_montgomery(BIGNUM *r, const BIGNUM *a, const BIGNUM *b,
                           BN_MONT_CTX *mont, BN_CTX *ctx)
+{
+    int ret = bn_mul_mont_fixed_top(r, a, b, mont, ctx);
+
+    bn_correct_top(r);
+    bn_check_top(r);
+
+    return ret;
+}
+
+int bn_mul_mont_fixed_top(BIGNUM *r, const BIGNUM *a, const BIGNUM *b,
+                          BN_MONT_CTX *mont, BN_CTX *ctx)
 {
     BIGNUM *tmp;
     int ret = 0;
-#if defined(OPENSSL_BN_ASM_MONT) && defined(MONT_WORD)
     int num = mont->N.top;
 
+#if defined(OPENSSL_BN_ASM_MONT) && defined(MONT_WORD)
     if (num > 1 && a->top == num && b->top == num) {
         if (bn_wexpand(r, num) == NULL)
             return 0;
         if (bn_mul_mont(r->d, a->d, b->d, mont->N.d, mont->n0, num)) {
             r->neg = a->neg ^ b->neg;
             r->top = num;
-            bn_correct_top(r);
+            r->flags |= BN_FLG_FIXED_TOP;
             return 1;
         }
     }
 #endif
 
+    if ((a->top + b->top) > 2 * num)
```

Comment:
@dot-asm I am wondering why this would not break in cases where `a` and `b` are sufficiently zero padded? And in fact why it is necessary at all?

Reply:
The question if it's necessary is kind of misplaced, because I for one argued [argued against the check](https://github.com/openssl/openssl/issues/5785#issuecomment-377191261). There I also said that at the very least, if one chooses to keep it, it would appropriate to move as close to caller as possible, which was done here in 3c97e4121ecec20cfac433883cd4709580a05620.

As for first question, what does "sufficiently zero-padded" mean? If it means *excessively* zero-padded, then it would be equivalent to feeding arbitrary data, and it would deserve to fail.